File: frysk-imports/util/eio/cni/MmapByteBuffer.cxx
// From the VENUS project.  Copyright 2004, 2005, Andrew Cagney
// Licenced under the terms of the Eclipse Public Licence.
// Licenced under the terms of the GNU CLASSPATH Licence.

#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <alloca.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>

#include <gcj/cni.h>
#include "util/eio/Buffer.h"
#include "util/eio/ByteBuffer.h"
#include "util/eio/MmapByteBuffer.h"
#include "util/eio/MmapByteBuffer$Mmap.h"

jlong
util::eio::MmapByteBuffer$Mmap::mmap (jstring file, jlong length)
{
  int pathlength = JvGetStringUTFLength (file);
  char *pathname = (char *) alloca (pathlength + 1);
  JvGetStringUTFRegion (file, 0, pathlength, pathname);
  pathname[pathlength] = '\0';

  errno = 0;
  int fd = open (pathname, O_RDONLY);
  if (errno != 0) {
    ::perror ("open");
    ::exit (errno);
  }

  errno = 0;
  void* buffer = ::mmap (NULL, length, PROT_READ | PROT_WRITE, MAP_PRIVATE,
			 fd, 0);
  if (errno != 0) {
    ::perror ("mmap");
    ::exit (errno);
  }

  return (jlong) (long) buffer;
}

jint
util::eio::MmapByteBuffer::peek (jlong caret)
{
  u_int8_t *p = (u_int8_t*) (long) this->map->byteData;
  return p[caret];
}

void
util::eio::MmapByteBuffer::poke (jlong caret, jint value)
{
  u_int8_t *p = (u_int8_t*) (long) this->map->byteData;
  p[caret] = value;
}

jlong
util::eio::MmapByteBuffer::peek (jlong caret, jbyteArray bytes, jlong off,
			     jlong len)
{
  u_int8_t *p = (u_int8_t*) (long) this->map->byteData;
  memcpy (elements (bytes) + off, p + caret, len);
  return len;
}

File: frysk-imports/util/eio/MmapByteBuffer.java
// From the VENUS project.  Copyright 2004, 2005, Andrew Cagney
// Licenced under the terms of the Eclipse Public Licence.
// Licenced under the terms of the GNU CLASSPATH Licence.

package util.eio;

public class MmapByteBuffer
    extends ByteBuffer
{
    private static final class Mmap
    {
	static private native final long mmap (String name, long length);
	long capacity;
	long byteData;
	private Mmap (String name)
	{
	    java.io.File file = new java.io.File (name);
	    capacity = file.length ();
	    byteData = mmap (name, capacity);
	}
    }

    Mmap map;

    private MmapByteBuffer (Mmap map, long lowExtreem, long highExtreem)
    {
	super (lowExtreem, highExtreem);
	this.map = map;
    }

    private MmapByteBuffer (Mmap map)
    {
	this (map, 0, map.capacity);
    }

    public MmapByteBuffer (String name)
    {
	this (new Mmap (name));
    }

    public ByteBuffer slice (long off, long len)
    {
	return new MmapByteBuffer (map, lowWater + off,
				   lowWater + off + len);
    }


    native protected final int peek (long index);
    native protected final void poke (long index, int val);
    
    native protected final long peek (long index, byte[] bytes,
				      long off, long len);
}
